File: javaMonad/monad.java
package javaMonad;

import java.util.function.Function;

public class monad<T> {
    private T value;

    private monad(T value) {
        this.value = value;
    }

    public static <T> monad<T> unit(T value) {
        return new monad<>(value);
    }

    public <R> monad<R> bind(Function<T, monad<R>> func) {
        return func.apply(value);
    } 

    public T getValue() {
        return value;
    }

    public static void main(String[] args) {
        monad<Integer> valueMonad = monad.unit(2);
        monad<Integer> resultMonad = valueMonad.bind(value -> monad.unit(value + 1))
                                               .bind(value -> monad.unit(value * 2))
                                               .bind(value -> monad.unit(value - 2));

        System.out.println(resultMonad.getValue());
    }
}

File: php/monad.php
<?php

class Monad
{
    private $value;

    private function __construct($value)
    {
        $this->value = $value;
    }

    public static function unit($value)
    {
        return new self($value);
    }

    public function bind($func)
    {
        return $func($this->value);
    }

    public function getValue()
    {
        return $this->value;
    }
}

function incrementValue($value)
{
    return Monad::unit($value + 1);
}

function decrementValue($value)
{
    return Monad::unit($value - 1);
}
function doubleValue($value)
{
    return Monad::unit($value * 2);
}

$valueMonad = Monad::unit(2);
$resultMonad = $valueMonad
                ->bind('incrementValue')
                ->bind('doubleValue')
                ->bind('decrementValue');

echo $resultMonad->getValue();

?>
